File: year2022/day03/python/solution.py
import re

def main():
  print(part_1())
  print(part_2())


def part_1():
  sum = 0
  for rucksack in input:
    compartments = re.findall(f".{{1,{len(rucksack) // 2}}}", rucksack)
    sum += get_char_value(''.join(set(compartments[0]) & set(compartments[1])))

  return sum


def part_2():
  sum = 0
  groups = [input[i:i+3] + [None] * (3 - len(input[i:i+3])) for i in range(0, len(input), 3)]
  for group in groups:
    sum += get_char_value(''.join(set(group[0]) & set(group[1]) & set(group[2])))

  return sum


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [row.rstrip('\n') for row in file]


def get_char_value(c):
  return ord(c) - (ord(c) > 96 and 96 or 38)


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2022/day03/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

public class Solution {
  private static List<String> INPUT = new ArrayList<String>();

  public static void main(String[] args) {
    try {
      setup(Files.readAllLines(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

  public static Integer part1() {
    int sum = 0;
    for (String rucksack : INPUT) {
      String[] compartments = rucksack.split("(?<=\\G.{" + rucksack.length() / 2 + "})");
      sum += getCharValue((char) compartments[0].chars()
          .filter(c -> compartments[1].indexOf(c) != -1)
          .findFirst().orElseThrow());
    }
    return sum;
  }

  public static Integer part2() {
    String[][] groups = IntStream.range(0, INPUT.size() / 3)
        .mapToObj(i -> INPUT.subList(i * 3, (i + 1) * 3)
        .toArray(new String[0]))
        .toArray(String[][]::new);

    return Arrays.stream(groups)
        .mapToInt(group -> getCharValue((char) group[0].chars()
        .filter(c -> group[1].indexOf(c) != -1 && group[2].indexOf(c) != -1)
        .findFirst().orElseThrow())).sum();
  }

  public static void setup(List<String> input) {
    INPUT = input;
  }

  private static int getCharValue(char c) {
    return (int) (Character.isUpperCase(c) ? c - 38 : c - 96);
  }
}